### PR TITLE
[luci] Introduce PartitionMerge

### PR DESCRIPTION
**Pull request by seanshpark**
This will introduce PartitionMerge that provides merge_pgroups method to
merge PGroup in PGroups with certain conditions.

ONE-DCO-1.0-Signed-off-by: SaeHie Park <saehie.park@gmail.com>

### CODE REVIEW

**Comment by seanshpark** [comment]
For #6230, from draft #6231

**Comment by seanshpark** [comment]
Current condition of merge still has problems. Still work in progress...

File: compiler/luci/partition/src/PartitionMerge.cpp [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+
+#include "PartitionMerge.h"
+
+#include <algorithm>
+
+namespace
+{
+
+/**
+ * @brief return true of pgroup_i output is one of the inputs of pgroup
```
Comment (mhs4670go) [inline compiler/luci/partition/src/PartitionMerge.cpp:25]:
```suggestion
 * @brief return true if pgroup_i output is one of the inputs of pgroup
```

File: compiler/luci/partition/src/PartitionMerge.cpp [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+}
+
+/**
+ * @brief return true if there is only one input or all the inputs have same group
+ * @note  pgroups is used to find group of pgroup
+ */
+bool is_input_same(const luci::PGroup *pgroup, const luci::PGroups *pgroups)
+{
+  assert(pgroups != nullptr);
+  assert(pgroup != nullptr);
```
Comment (mhs4670go) [inline compiler/luci/partition/src/PartitionMerge.cpp:47]: maybe just returning false is better?

Comment (seanshpark) [reply compiler/luci/partition/src/PartitionMerge.cpp:47]: usually I add assert for the condition of the method. this assert is to check the condition. the caller should give correct condition.

File: compiler/luci/partition/src/PartitionMerge.cpp [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+  const luci::PGroup *input_pgroup = nullptr;
+  std::string group;
+  for (auto &input : pgroup->inputs)
+  {
+    auto input_group = pgroups->group_of(input);
+    // NOTE: all the nodes should be registered and return should be valid group.
+    // convert_to_proups() should ensure this.
+    // assert here to find if there is any problem with this.
+    assert(not input_group.empty());
+    if (input_group.empty())
```
Comment (mhs4670go) [inline compiler/luci/partition/src/PartitionMerge.cpp:58]: Maybe this `if` condition would never be called.

Comment (seanshpark) [reply compiler/luci/partition/src/PartitionMerge.cpp:58]: As added a comment, this is a guard to find if any problems.
`if` itself is for release build.

File: compiler/luci/partition/src/PartitionMerge.cpp [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+/**
+ * @brief merge pgroup into pgroup_i
+ * @note  output of pgroup_i should be input of pgroup
+ */
+void merge_into(luci::PGroup *pgroup, luci::PGroup *pgroup_i)
+{
+  for (auto &pnode : pgroup->pnodes)
+  {
+    // update pgroup for this pnode
+    pnode->pgroup = pgroup_i;
```
Comment (mhs4670go) [inline compiler/luci/partition/src/PartitionMerge.cpp:95]: Just curiosity, do they have to have the same group?

Comment (seanshpark) [reply compiler/luci/partition/src/PartitionMerge.cpp:95]: No. I the beginning, every pnode(which holds node) reside in it's own pgroup. that is one pnode per one pgroup.
`Merge` is to merge two pgroup if condition is met.

**Review by mhs4670go — APPROVED** [review]
LGTM

**Pull request merged**